scripts/travis/build-all-in-one-image.sh

```diff
@@ -26,42 +29,89 @@ run_integration_test() {
-  if [[ ("$BRANCH" == "master" || $BRANCH =~ ^v[0-9]+\.[0-9]+\.[0-9]+$) && "$TRAVIS_SECURE_ENV_VARS" == "true" ]]; then
-    echo 'upload to Docker Hub'
+docker_buildx() {
+  CMD_ROOT=${1}
+  FLAGS=${@:2}
+
+  if [[ "$FLAGS" == *"-push"*  ]]; then
+    if [[ ("$BRANCH" == "master" || $BRANCH =~ ^v[0-9]+\.[0-9]+\.[0-9]+$) && "$TRAVIS_SECURE_ENV_VARS" == "true" ]]; then
+      echo 'upload to Docker Hub'
+    else
+      echo 'skip docker upload for PR'
+      exit 0
+    fi
+  fi
+
+  docker buildx build -f ${CMD_ROOT}/Dockerfile ${FLAGS} ${CMD_ROOT}
+}
+
+image_tags_for() {
+  REPO=${1}
+
+  major=""
+  minor=""
+  patch=""
+
+  if [[ "$BRANCH" == "master" ]]; then
+    TAG="latest"
+  elif [[ $BRANCH =~ ^v([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]; then
+    major="${BASH_REMATCH[1]}"
+    minor="${BASH_REMATCH[2]}"
+    patch="${BASH_REMATCH[3]}"
+    TAG=${major}.${minor}.${patch}
   else
-    echo 'skip docker upload for PR'
-    exit 0
+    TAG="${BRANCH///}"
+  fi
+
+  IMAGE_TAGS="--tag ${REPO}:${TAG}"
+
+  # add major, major.minor and major.minor.patch tags
+  if [[ -n $major ]]; then
+    IMAGE_TAGS="${IMAGE_TAGS} -t ${REPO}:${major}"
+    if [[ -n $minor ]]; then
+       IMAGE_TAGS="${IMAGE_TAGS} -t ${REPO}:${major}.${minor}"
+      if [[ -n $patch ]]; then
+          IMAGE_TAGS="${IMAGE_TAGS} -t ${REPO}:${major}.${minor}.${patch}"
+      fi
+    fi
   fi
-  export REPO=$1
-  bash ./scripts/travis/upload-to-docker.sh
+
+  echo "${IMAGE_TAGS}"
+}
+
+target_platforms() {
+  PLATFORMS=""
+  for arch in ${ARCHS}; do
+    if [ -n "${PLATFORMS}" ]; then
+      PLATFORMS="${PLATFORMS},linux/${arch}"
+    else
+      PLATFORMS="linux/${arch}"
+    fi
+  done
+  echo ${PLATFORMS}
 }
 
-make build-all-in-one GOOS=linux GOARCH=$GOARCH
-repo=jaegertracing/all-in-one
-docker build -f cmd/all-in-one/Dockerfile \
-        --target release \
-        --tag $repo:latest cmd/all-in-one \
-        --build-arg base_image=localhost/baseimg:1.0.0-alpine-3.12 \
-        --build-arg debug_image=localhost/debugimg:1.0.0-golang-1.15-alpine \
-        --build-arg TARGETARCH=$GOARCH
-run_integration_test $repo
-upload_to_docker $repo
-
-make build-all-in-one-debug GOOS=linux GOARCH=$GOARCH
-repo=jaegertracing/all-in-one-debug
-docker build -f cmd/all-in-one/Dockerfile \
-        --target debug \
-        --tag $repo:latest cmd/all-in-one \
-        --build-arg base_image=localhost/baseimg:1.0.0-alpine-3.12 \
-        --build-arg debug_image=localhost/debugimg:1.0.0-golang-1.15-alpine \
-        --build-arg TARGETARCH=$GOARCH
-run_integration_test $repo
-upload_to_docker $repo
-
-make build-otel-all-in-one GOOS=linux GOARCH=$GOARCH
-repo=jaegertracing/opentelemetry-all-in-one
-docker build -f cmd/opentelemetry/cmd/all-in-one/Dockerfile -t $repo:latest cmd/opentelemetry/cmd/all-in-one --build-arg TARGETARCH=$GOARCH
-run_integration_test $repo
-upload_to_docker $repo
+
+for arch in ${ARCHS}; do
+  make build-all-in-one GOOS=linux GOARCH=${arch}
```

Comment:
What's the difference in the clock time for this step with all arches vs. only one arch?

Reply:
we need to build each go binary for each arch.

the final Dockerfile not includes build stage, just copy binary from the step.

using `for in` here, only want to make ARCHS be easier to configurate

Reply:
My question is mainly to determine if we have a penalty in build time for CI. If it takes 5m to build *each* arch, then we'll have to rethink the approach. If it takes 5m for the first plus 1m for each subsequent arch, then it's fine.

Reply:
it will not take 5m for each arch. 

1. go build for amd64, go download deps (3m) + go prebuild common (0.5m) + go prebuild for amd64 (0.5m) + go build bin (1m)
2. go build for arm64, go prebuild for arm64 (0.5m) + go build bin (1m)
2. go build for s390x, go prebuild for s390x (0.5m) + go build bin (1m)

if we split, it will take 5m for each arch. this why i upgrade for jaeger-operator.

Reply:
Great, thanks for the confirmation. I wasn't sure whether the download step would be shared across steps. If it adds 90s for each subsequent arch, I think it's reasonable.